Added noexcept(false) to module destructors for gcc 6

Dr15Jones opened the pull request:
For now, we will continue to allow module's destructors to throw exceptions. Explicitly added 'noexcept(false)' to base class destructors to silence warnings from gcc 6.

Dr15Jones committed:
Added noexcept(false) to module destructors for gcc 6

For now, we will continue to allow module's destructors to throw exceptions. Explicitly added 'noexcept(false)' to base class destructors to silence warnings from gcc 6.

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for CMSSW_8_1_X.

It involves the following packages:

FWCore/Framework
FWCore/Integration
FWCore/Sources


@cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit, @wmtan this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

cmsbuild commented:
-1

Tested at: 83f28029269fa42e13b93a68da2489ce7846cd6d

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14396/12852/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Validation/EventGenerator/src/DQMHelper.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Validation/EventGenerator/src/TauDecay.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Validation/EventGenerator/src/TauDecay_CMSSW.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Validation/EventGenerator/src/CaloCellId.cc 
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Validation/RecoParticleFlow/plugins/GenericBenchmarkAnalyzer.cc:1:0:
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Validation/RecoParticleFlow/plugins/GenericBenchmarkAnalyzer.h:21:11: error: looser throw specifier for 'virtual GenericBenchmarkAnalyzer::~GenericBenchmarkAnalyzer() noexcept (false)'
   virtual ~GenericBenchmarkAnalyzer();
           ^
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Validation/RecoParticleFlow/plugins/GenericBenchmarkAnalyzer.h:11:0,
                 from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Validation/RecoParticleFlow/plugins/GenericBenchmarkAnalyzer.cc:1:
/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-05-06-1100/src/RecoParticleFlow/Benchmark/interface/GenericBenchmark.h:29:11: error:   overriding 'virtual GenericBenchmark::~GenericBenchmark() noexcept'
</pre>

Dr15Jones committed:
Added noexcept(false) to GenericBenchmark's destructor

Because a module uses dual inheritance to GenericBenchmark and the base class for modules' use noexcept(false) we had to make GenericBenchmark's destructor just as loose.

cmsbuild commented:
Pull request #14396 was updated. @smuzaffar, @Dr15Jones, @cvuosalo, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

cmsbuild commented:
-1

Tested at: 83a20874b2d00e0275f5259ed2582bc669815649

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14396/12856/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Alignment/SurveyAnalysis/plugins/SurveyDBUploader.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Alignment/SurveyAnalysis/plugins/SurveyInputTrackerFromDB.cc 
Entering library rule at src/Alignment/LaserAlignment/plugins
Entering library rule at SimCalorimetry/CastorSim
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Alignment/LaserAlignmentSimulation/test/SimAnalyzer.cc:9:0:
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-05-06-1100/src/Alignment/LaserAlignmentSimulation/test/SimAnalyzer.h:31:3: error: looser throw specifier for 'virtual SimAnalyzer::~SimAnalyzer() noexcept (false)'
   ~SimAnalyzer();
   ^
In file included from /cvmfs/cms-ib.cern.ch/2016-19/slc6_amd64_gcc530/lcg/root/6.06.00-ocfgfb/include/TNamed.h:26:0,
                 from /cvmfs/cms-ib.cern.ch/2016-19/slc6_amd64_gcc530/lcg/root/6.06.00-ocfgfb/include/TAxis.h:25,
                 from /cvmfs/cms-ib.cern.ch/2016-19/slc6_amd64_gcc530/lcg/root/6.06.00-ocfgfb/include/TH1.h:25,
</pre>

slava77 commented:
@Dr15Jones 
It looks like more changes are necessary here.
Is there an issue to pick the compile errors locally?

Dr15Jones committed:
Removed unnecessary inheritance on TObject

The EDProducer did not need to inherit from TObject and the inheritance was interfering with the change of the destructor to be noexcept(false).

Dr15Jones committed:
Explicitly allow destructor to throw exceptions (for gcc6)

Dr15Jones committed:
Explicitly allow destructor to throw exceptions

This class is used as a base class for multiple inheritance with a
framework module base class. For gcc 6 the framework modules explicitly
declared that the destructors are allowed to thow exceptions [noexcept(false)].
However, in the case of multiple inheritance, this meant all other base classes
used also need to allow exceptions else there is a compile time error.

cmsbuild commented:
Pull request #14396 was updated. @ghellwig, @smuzaffar, @civanch, @Dr15Jones, @cerminar, @cvuosalo, @perrozzi, @mdhildreth, @dmitrijus, @cmsbuild, @alja, @franzoni, @covarell, @thuer, @deguio, @slava77, @ggovi, @mmusich, @vanbesien, @govoni, @davidlange6 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
+1

civanch commented:
+1

dmitrijus commented:
+1

cmsbuild commented:
-1

Tested at: 1e89dbc69790462d865ea1186cad8663d0496ab4

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14396/12900/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestFWCoreIntegrationStandalone had ERRORS

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14396/12900/summary.html

Dr15Jones commented:
The test failure appears to be a transient problem. From the logs it says
```
===== Test "TestFWCoreIntegrationStandalone" ====
Running .

Traceback helper thread failed to read from parent: Bad file descriptor
/bin/sh: line 1: 26777 Aborted                 TestFWCoreIntegrationStandalone

---> test TestFWCoreIntegrationStandalone had ERRORS
```

When I run the tests in my local area they pass fine.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

ghellwig commented:
+1

cmsbuild commented:
+1
Tested at: 1e89dbc69790462d865ea1186cad8663d0496ab4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14396/12904/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14396/12904/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/14396	1e89dbc
- changes are as described
- jenkins tests pass

davidlange6 commented:
+1
